selfdrive/car/nissan/interface.py

```diff
@@ -0,0 +1,199 @@
+    self.acc_active_prev = 0
+    self.gas_pressed_prev = False
+
+    # *** init the major players ***
+    self.CS = CarState(CP)
+    self.VM = VehicleModel(CP)
+    self.pt_cp = get_powertrain_can_parser(CP)
+    self.adas_cp = get_adas_can_parser(CP)
+    self.cam_cp = get_camera_can_parser(CP)
+
+    self.gas_pressed_prev = False
+    self.brake_pressed_prev = False
+
+    self.CC = None
+    if CarController is not None:
+      self.CC = CarController(CP.carFingerprint)
+
+  @staticmethod
+  def compute_gb(accel, speed):
+    return float(accel) / 4.0
+
+  @staticmethod
+  def get_params(candidate, fingerprint=gen_empty_fingerprint(), has_relay=False, car_fw=[]):
+    ret = car.CarParams.new_message()
+
+    ret.carName = "nissan"
+    ret.carFingerprint = candidate
+    ret.isPandaBlack = has_relay
+    ret.safetyModel = car.CarParams.SafetyModel.nissan
+
+    ret.enableCruise = True
+    ret.steerLimitAlert = False
+
+    ret.enableCamera = True
+
+    ret.steerRateCost = 0.5
+
+    if candidate in [CAR.XTRAIL]:
+      ret.mass = 1610 + STD_CARGO_KG
+      ret.wheelbase = 2.705
+      ret.centerToFront = ret.wheelbase * 0.44
+      ret.steerRatio = 17
+      ret.steerActuatorDelay = 0.1
+      ret.lateralTuning.pid.kf = 0.00006
+      ret.lateralTuning.pid.kiBP, ret.lateralTuning.pid.kpBP = [[0.0], [0.0]]
+      ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[0.01], [0.005]]
+      ret.steerMaxBP = [0.] # m/s
+      ret.steerMaxV = [1.]
+
+    ret.steerControlType = car.CarParams.SteerControlType.angle
+    ret.steerRatioRear = 0.
+    # testing tuning
+
+    # No long control in nissan
+    ret.gasMaxBP = [0.]
+    ret.gasMaxV = [0.]
+    ret.brakeMaxBP = [0.]
+    ret.brakeMaxV = [0.]
+    ret.longitudinalTuning.deadzoneBP = [0.]
+    ret.longitudinalTuning.deadzoneV = [0.]
+    ret.longitudinalTuning.kpBP = [0.]
+    ret.longitudinalTuning.kpV = [0.]
+    ret.longitudinalTuning.kiBP = [0.]
+    ret.longitudinalTuning.kiV = [0.]
+
+    # end from gm
+
+    # TODO: get actual value, for now starting with reasonable value for
+    # civic and scaling by mass and wheelbase
+    ret.rotationalInertia = scale_rot_inertia(ret.mass, ret.wheelbase)
+
+    # TODO: start from empirically derived lateral slip stiffness for the civic and scale by
+    # mass and CG position, so all cars will have approximately similar dyn behaviors
+    ret.tireStiffnessFront, ret.tireStiffnessRear = scale_tire_stiffness(ret.mass, ret.wheelbase, ret.centerToFront)
+
+    return ret
+
+  # returns a car.CarState
+  def update(self, c, can_strings):
+    self.pt_cp.update_strings(can_strings)
+    self.cam_cp.update_strings(can_strings)
+    self.adas_cp.update_strings(can_strings)
+
+    self.CS.update(self.pt_cp, self.adas_cp, self.cam_cp)
+
+    # create message
+    ret = car.CarState.new_message()
+
+    ret.canValid = self.pt_cp.can_valid and self.adas_cp.can_valid and self.cam_cp.can_valid
+
+    # speeds
+    ret.vEgo = self.CS.v_ego
```

Comment:
Recently made improvement in car abstraction (see other cars)
all these copies are now obsolete: CS should just return the capnp struct directly.